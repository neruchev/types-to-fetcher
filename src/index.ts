import axios, { AxiosError } from 'axios';
import { compile } from 'path-to-regexp';

export type Payload = {
  Body?: unknown;
  Querystring?: unknown;
  Params?: Record<string, string>;
  Headers?: unknown;
  Reply: unknown;
};

export type Options<Config extends Payload, Error> = {
  baseURL: string;
  effect?: Effect<Config, Error>;
};

export type UnionToIntersection<U> = (
  U extends any ? (k: U) => void : never
) extends (k: infer I) => void
  ? I
  : never;

export type LastOf<T> = UnionToIntersection<
  T extends any ? () => T : never
> extends () => infer R
  ? R
  : never;

export type TuplifyUnion<
  T,
  L = LastOf<T>,
  N = [T] extends [never] ? true : false
> = true extends N ? [] : [...TuplifyUnion<Exclude<T, L>>, L];

export type Effect<Config extends Payload, Error> = (
  action: Fetcher<Config, Error>,
  params: {
    baseURL: string;
    endpoint: string;
    method: string;
  }
) => Fetcher<Config, Error>;

export type EffectAbort = {
  abort: AbortController['abort'];
};

export type Schema<API> = {
  [Endpoint in keyof API]: TuplifyUnion<keyof API[Endpoint]>;
};

export type Methods<MethodsRecord extends object, Error> = {
  [Method in keyof MethodsRecord]: Fetcher<MethodsRecord[Method], Error> &
    EffectAbort;
};

export type Endpoints<EndpointsRecord extends object, Error> = {
  [Endpoint in keyof EndpointsRecord]: Methods<
    EndpointsRecord[Endpoint],
    Error
  >;
};

export type Fetcher<Config extends Payload, Error> = (
  data: Omit<Config, 'Reply' | 'Headers'>
) => Promise<Exclude<Config['Reply'], Error>>;

export const fetcher = <Config extends Payload, Error>(
  baseURL: string,
  url: string,
  method: string
): Fetcher<Config, Error> & EffectAbort => {
  let abortController: AbortController | undefined;

  return Object.assign(
    async ({
      Body,
      Querystring,
      Params,
    }: Partial<{
      Body: Config['Body'];
      Querystring: Config['Querystring'];
      Params: Config['Params'];
    }>) => {
      abortController = new AbortController();

      try {
        const { data } = await axios({
          url: compile(url)(Params),
          method,
          baseURL,
          data: Body,
          params: Querystring,
          withCredentials: true,
          signal: abortController.signal,
        });

        return data;
      } catch (error) {
        const { response, message, code } = error as AxiosError<Error>;

        if (code === 'ERR_CANCELED') {
          return;
        }

        throw (response?.data as any)?.error ?? message ?? 'Unknown error';
      }
    },
    {
      abort: () => {
        if (!abortController) {
          return;
        }

        abortController.abort();
      },
    }
  );
};

export const makeApi = <API extends object, Error, Effect = void>(
  schema: Schema<API>,
  options: Options<Payload, Error>
): Effect extends void ? Endpoints<API, Error> : Effect => {
  const result = schema as any;
  const { baseURL, effect } = options;

  for (const endpoint in result) {
    result[endpoint] = result[endpoint].reduce(
      (acc: Record<string, Fetcher<Payload, Error>>, method: string) => {
        const handler = fetcher(baseURL, endpoint, method);
        const abort = handler.abort;

        acc[method] = effect
          ? Object.assign(effect(handler, { endpoint, method, baseURL }), {
              abort,
            })
          : handler;

        return acc;
      },
      {}
    );
  }

  return result;
};
