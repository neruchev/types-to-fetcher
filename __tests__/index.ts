import { fetcher, makeApi } from '../src';
import axios from 'axios';

jest.mock('axios');
const mocked = jest.mocked(axios, { shallow: true });

mocked.mockImplementation((config: any) => {
  if (config?.onUploadProgress) {
    config.onUploadProgress({ loaded: 50, total: 100 });
  }
  if (config?.onDownloadProgress) {
    config.onDownloadProgress({ loaded: 50, total: 100 });
  }

  return Promise.resolve({ data: { foo: 'bar' } }) as never;
});

interface API {
  '/': {
    GET: {
      Reply: {
        version: string;
        mode: 'production' | 'development';
      };
    };
  };
  '/foo/:bar': {
    GET: {
      Params: { bar: string };
      Reply: string;
    };
    POST: {
      Params: { bar: string };
      Body: { baz: string };
      Reply: Error | string;
    };
  };
}

describe('Fetcher', () => {
  test('Fetcher call returns a function', () => {
    expect(typeof fetcher('', '', '')).toBe('function');
    expect(typeof fetcher('https://example.com', '/', 'GET')).toBe('function');
  });

  test("Url without format expression won't compile", () => {
    const { signal } = new AbortController();

    fetcher('', '/x', 'GET')({ Axios: { signal } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'GET',
      url: '/x',
      withCredentials: true,
      signal,
    });
  });

  test('Compiling a url with a format expression will return the compiled url', () => {
    const { signal } = new AbortController();

    fetcher('', '/x/:y', 'GET')({ Params: { y: 'z' }, Axios: { signal } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'GET',
      url: '/x/z',
      withCredentials: true,
      signal,
    });
  });

  test('The Body forwards correctly', () => {
    const { signal } = new AbortController();

    fetcher('', '/', 'POST')({ Body: { a: 'b' }, Axios: { signal } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'POST',
      url: '/',
      withCredentials: true,
      data: { a: 'b' },
      signal,
    });
  });

  test('The Querystring forwards correctly', () => {
    const { signal } = new AbortController();

    fetcher('', '/', 'GET')({ Querystring: { a: 'b' }, Axios: { signal } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: '',
      method: 'GET',
      url: '/',
      withCredentials: true,
      params: { a: 'b' },
      signal,
    });
  });

  test('Response returned correctly', async () => {
    const { signal } = new AbortController();

    const response = await fetcher('', '/', 'GET')({ Axios: { signal } });

    expect(response).toEqual({ foo: 'bar' });
  });

  test('Signal exist, work correctly', async () => {
    const { signal } = new AbortController();

    const response = await fetcher('', '/', 'GET')({ Axios: { signal } });

    expect(axios).toHaveBeenCalledWith(expect.objectContaining({ signal }));
    expect(response).toEqual({ foo: 'bar' });
  });

  test('onUploadProgress is called during upload', async () => {
    const onUploadProgressMock = jest.fn();

    await fetcher(
      'https://example.com',
      '/upload',
      'POST'
    )({
      Body: { some: 'data' },
      Axios: { onUploadProgress: onUploadProgressMock },
    });

    expect(onUploadProgressMock).toHaveBeenCalled();
  });

  test('onDownloadProgress is called during download', async () => {
    const onDownloadProgressMock = jest.fn();

    await fetcher(
      'https://example.com',
      '/download',
      'GET'
    )({
      Axios: { onDownloadProgress: onDownloadProgressMock },
    });

    expect(onDownloadProgressMock).toHaveBeenCalled();
  });
});

describe('Make API', () => {
  test('The generator output snapshot is correctly', () => {
    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      { baseURL: '/x' }
    );

    expect(api).toMatchSnapshot();
  });

  test('The effect forwards correctly', async () => {
    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      {
        baseURL: 'hrrps://api.mysite.com/',
        timeout: 20000,
        effect: (_, params) => ({ params } as never),
      }
    );

    expect(api).toMatchSnapshot();
  });

  test('The baseURL forwards correctly', () => {
    const { signal } = new AbortController();

    const api = makeApi<API, {}>(
      {
        '/': ['GET'],
        '/foo/:bar': ['GET', 'POST'],
      },
      { baseURL: 'hrrps://api.mysite.com/' }
    );

    api['/'].GET({ Axios: { signal } });

    expect(mocked).toHaveBeenCalledWith({
      baseURL: 'hrrps://api.mysite.com/',
      method: 'GET',
      url: '/',
      withCredentials: true,
      signal,
    });
  });
});
