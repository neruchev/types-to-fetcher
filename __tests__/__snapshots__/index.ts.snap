// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make API The effect forwards correctly 1`] = `
{
  "/": {
    "GET": {
      "params": {
        "baseURL": "hrrps://api.mysite.com/",
        "endpoint": "/",
        "method": "GET",
        "paramsSerializer": undefined,
        "timeout": 20000,
      },
    },
  },
  "/foo/:bar": {
    "GET": {
      "params": {
        "baseURL": "hrrps://api.mysite.com/",
        "endpoint": "/foo/:bar",
        "method": "GET",
        "paramsSerializer": undefined,
        "timeout": 20000,
      },
    },
    "POST": {
      "params": {
        "baseURL": "hrrps://api.mysite.com/",
        "endpoint": "/foo/:bar",
        "method": "POST",
        "paramsSerializer": undefined,
        "timeout": 20000,
      },
    },
  },
}
`;

exports[`Make API The generator output snapshot is correctly 1`] = `
{
  "/": {
    "GET": [Function],
  },
  "/foo/:bar": {
    "GET": [Function],
    "POST": [Function],
  },
}
`;
